File: medium/0024_swap_nodes_in_pairs/solution.cpp
#include "../../utils/cpp/nodes.hpp"

class Solution {
public:
  utils::ListNode *swapPairs(utils::ListNode *head) {
    if (!head)
      return nullptr;

    utils::ListNode *dummy = new utils::ListNode(-1, head);

    utils::ListNode *prev = dummy;
    utils::ListNode *curr = head;
    utils::ListNode *temp;

    while (curr) {
      if (curr->next) {
        temp = curr->next;
        prev->next = temp;
        curr->next = temp->next;
        temp->next = curr;
      }

      prev = curr;
      curr = curr->next;
    }

    return dummy->next;
  }
};

File: medium/0024_swap_nodes_in_pairs/solution.py
from typing import Optional

from utils.python.nodes import ListNode


class Solution:
    def swapPairs(self, head: Optional[ListNode]) -> Optional[ListNode]:
        if not head:
            return head

        dummy = ListNode(-1)
        dummy.next = head

        prev = dummy
        curr = head

        while curr:
            if curr.next:
                temp = curr.next
                prev.next, curr.next, temp.next = temp, temp.next, curr

            prev = curr
            curr = curr.next

        return dummy.next
